### ISSUE TITLE
DataProvider and() is confusing since it produces a NEW DataProvider

### ISSUE THREAD

**Issue by mvysny** [mvysny/vaadin-on-kotlin#18]
It's not clear from the `and()` method name that a new DP is produced by the call. That is also the case of `configurableFilter()`.

Perhaps the data providers offered from the entities should be configurable right from the start. That makes sense since then filter components can directly set filters to those data providers, without having to call the confusing `configurableFilter()` method. That would also allow us to call `setFilter()` directly which makes sense.

When a function creates a new instance, it is typically prefixed with `with`. Hence the `configurableFilter()` should be renamed to `withConfigurableFilter()`. Also, the `and()` should be renamed to `withFilter()`.

With these changes, current use-cases of creating a grid with filtered items are still valid, for example:

```kotlin
personGrid = grid(Person::class, dataProvider = Person.dataProvider) {
  // since dataProvider is configurable, the generateFilterComponents() function will work properly
}
```

also

```kotlin
personGrid = grid(Person::class, dataProvider = Person.dataProvider.withFilter { Person::age between 20..60 })
```

or

```kotlin
personGrid = grid(Person::class, dataProvider = Person.dataProvider.apply { setFilter { Person::age between 20..60 }})
```

However, there is now a problem that the generated filter bar will silently overwrite the initial filter set by the programmer ...

**Comment by mvysny** [comment]
One way would be to call `dataProvider = Person.dataProvider.withFilter { Person::age between 20..60 } .withFilter {}` but that's error-prone since there is no check that the programmer has called that...

Aha! If the semantics of `withFilter()` would be that it creates a new data provider with given unremovable filter and additional support of setting additional filters, that would help.

**Comment by mvysny** [comment]
OK, removed `and` and introduced `withFilter` and `setFilter`; now generated filters won't overwrite the filter set in stone by the programmer using the `withFilter`. Much better :)

Fixed in VoK 0.4.3
